Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift

```diff
@@ -118,11 +116,24 @@ extension BTThreeDSecureV2Provider: CardinalValidationDelegate {
         stepUpValidated validateResponse: CardinalResponse!,
         serverJWT: String!
     ) {
-        let actionCodeString = analyticsString(for: validateResponse.actionCode)
-        apiClient.sendAnalyticsEvent("ios.three-d-secure.verification-flow.cardinal-sdk.action-code.\(actionCodeString)")
-
         switch validateResponse.actionCode {
-        case .success, .noAction, .failure:
+        case .success:
+            apiClient.sendAnalyticsEvent(BTThreeDSecureAnalytics.challengeSucceeded)
+            BTThreeDSecureAuthenticateJWT.authenticate(
+                jwt: serverJWT,
+                withAPIClient: apiClient,
+                forResult: lookupResult,
+                completion: completionHandler
+            )
+        case .failure:
+            apiClient.sendAnalyticsEvent(BTThreeDSecureAnalytics.challengeFailed)
```

Comment:
Is ".failure" from Cardinal considered "challengeFailed". The` challengeFailed` analytic description in our Excel doc says "Fired when the Cardinal SDK returns an error in completing the 3DS challenge". Both are `challengeFailed`, right?

Reply:
So for analytics, we don't distinguish service level error and error based on user's input of credentials?

Reply:
Thank you for looking those up. I will do that first in the future.

Reply:
I wouldn't change any business logic for the purpose of analytics. [We should keep it like we had it](https://github.com/braintree/braintree_ios/blob/5.x/Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.m#L104-L113), and continue to call `authenticate` for .success, .noAction, and .failure

Reply:
The `3ds:verify:challenge:failed` event should be fired [here](https://github.com/braintree/braintree_ios/blob/02234df74c9837637165c668b44f879fbb945f4c/Sources/BraintreeThreeDSecure/BTThreeDSecureV2Provider.swift#L128-L141).

Reply:
➕ that makes sense to me!

Reply:
I think we didn't send `challenge:failed` for error other than cancel https://github.com/braintree/braintree_ios/blob/9cb3aa0b50b027e63f93a8abc4cda6492d897b1a/Sources/BraintreeSEPADirectDebit/BTSEPADirectDebitClient.swift#L190

Reply:
@KunJeongPark  - Making sure [this comment](https://github.com/braintree/braintree_ios/pull/1025/files#r1184279390) gets addressed. We want to revert to the original business logic.

Reply:
you mean sending challenge fail for error, timeout block? It's done. we are calling authenticate for all the cases, I just separated them out to send different analytic messages.

Reply:
👋 @KunJeongPark my [comments](https://github.com/braintree/braintree_ios/pull/1025/files#r1184279390) here still haven't been addressed.

Reply:
@KunJeongPark I would consolidate where possible, so:

```
        case .success, .noAction:
            apiClient.sendAnalyticsEvent(BTThreeDSecureAnalytics.challengeSucceeded)
            BTThreeDSecureAuthenticateJWT.authenticate(
                jwt: serverJWT,
                withAPIClient: apiClient,
                forResult: lookupResult,
                completion: completionHandler
            )
        case .failure:
            apiClient.sendAnalyticsEvent(BTThreeDSecureAnalytics.challengeFailed)
            BTThreeDSecureAuthenticateJWT.authenticate(
                jwt: serverJWT,
                withAPIClient: apiClient,
                forResult: lookupResult,
                completion: completionHandler
            )
```

Otherwise, LGTM!

Reply:
We could also go a step further to avoid duplicating the authenticate() call with an `if` inside the triple case block. Not sure which would be the most readable.